### PR TITLE
better session handling

### PR DESCRIPTION
**Pull request by lavrukov**
If we received a TLI or BadSession inside a transaction, this means that the transaction was invalidated on the YDB side and sending commit() or rollback() is incorrect, since in this case they always generate a BadSession.

In order to write tests for this behavior that will work both on the real YDB and on the mock, we have to add a check for read locks inside the InMemoryRepository.

### CODE REVIEW

File: repository-ydb-v1/src/main/java/tech/ydb/yoj/repository/ydb/YdbRepositoryTransaction.java [hunk 1]
```diff
@@ -136,8 +138,8 @@ public void rollback() {
     private void doCommit() {
         try {
             Status status = YdbOperations.safeJoin(session.commitTransaction(txId, new CommitTxSettings()));
-            validatePkConstraint(status.getIssues());
             validate("commit", status.getCode(), status.toString());
```
Comment (lavrukov) [inline repository-ydb-v1/src/main/java/tech/ydb/yoj/repository/ydb/YdbRepositoryTransaction.java:141]: It looks like validate() should be called immediately after safeJoin()

File: repository-ydb-v1/src/test/java/tech/ydb/yoj/repository/ydb/YdbRepositoryIntegrationTest.java [hunk 2]
```diff
@@ -863,6 +866,30 @@ public void schemaWithHint() {
         repository.schema(HintAutoPartitioningByLoad.class).create();
     }
 
+
```
Comment (nvamelichev) [inline repository-ydb-v1/src/test/java/tech/ydb/yoj/repository/ydb/YdbRepositoryIntegrationTest.java:869]: minor: style: excessive newline

File: repository-ydb-v1/src/test/java/tech/ydb/yoj/repository/ydb/YdbRepositoryIntegrationTest.java [hunk 3]
```diff
@@ -863,6 +866,30 @@ public void schemaWithHint() {
...
+        );
+
+        tx.table(Project.class).find(id2);
+
+        db.tx(() -> db.projects().save(new Project(id2, "name2")));
+
+        tx.table(Project.class).save(new Project(id1, "name1")); // make tx available for TLI
+
+        assertThatExceptionOfType(OptimisticLockException.class)
+                .isThrownBy(() -> tx.table(Project.class).find(id2));
```
Comment (nvamelichev) [inline repository-ydb-v1/src/test/java/tech/ydb/yoj/repository/ydb/YdbRepositoryIntegrationTest.java:888]: Modern YDB versions can now return OptimisticLock from statements, it that correct?

File: repository-ydb-v1/src/test/java/tech/ydb/yoj/repository/ydb/YdbRepositoryIntegrationTest.java [hunk 4]
```diff
@@ -863,6 +866,30 @@ public void schemaWithHint() {
...
+
+        tx.commit(); // Commit shouldn't throw an BadSession or other exceptions
```
Comment (nvamelichev) [inline repository-ydb-v1/src/test/java/tech/ydb/yoj/repository/ydb/YdbRepositoryIntegrationTest.java:890]: Is it even correct to call `commit()` on a transaction *that got invalidated* (allowing no-error `rollback()` I can kinda understand)? 
I think trying to commit an invalidated transaction should definitely cause `IllegalStateException` or something like that, no?

**Review by nvamelichev — CHANGES_REQUESTED** [review]

File: repository-ydb-v1/src/test/java/tech/ydb/yoj/repository/ydb/YdbRepositoryIntegrationTest.java (hunk 3 above)
Comment (lavrukov) [reply repository-ydb-v1/src/test/java/tech/ydb/yoj/repository/ydb/YdbRepositoryIntegrationTest.java:888]: Yes, this test works over latest docker image (23.3).

File: repository-ydb-v1/src/test/java/tech/ydb/yoj/repository/ydb/YdbRepositoryIntegrationTest.java (hunk 4 above)
Comment (lavrukov) [reply repository-ydb-v1/src/test/java/tech/ydb/yoj/repository/ydb/YdbRepositoryIntegrationTest.java:890]: In this PR, I wanted to touch the old logic to a minimum. commit() is controlled from TxManger, it cannot make a commit after TLI or BadSession, but maybe someone other can. To reduce the likelihood of strange illegals falling out, I did not check it here. This is what this test shows.
But still, it’s really more correct to throw an exception here, I’ll do it.

Also I have a proposal-PR which make works with session and exceptions more understandable (maybe), but I pull it after this.

**Review by nvamelichev — APPROVED** [review]
LGTM :+1:

**Comment by nvamelichev** [comment]
I will look at the in-memory changes more thoroughly on Monday, January 8. New locking logic seems a bit strange, e.g. you can get a writable tx shard that (seemingly) never checks for TLI

**Comment by nvamelichev** [comment]
> Looks like write operations can't return TLI right now. I showed it in test writeDontProduceTLI()

Are we sure that this is the case always, for recent versions of YDB? Maybe it's better to be cautious and raise TLI here even if YDB currently does not.

**Comment by lavrukov** [comment]
> Are we sure that this is the case always, for recent versions of YDB? Maybe it's better to be cautious and raise TLI here even if YDB currently does not.

We don't have any quaranty

**Comment by lavrukov** [comment]
> Are we sure that this is the case always, for recent versions of YDB? Maybe it's better to be cautious and raise TLI here even if YDB currently does not.

We have no any guarantees about future, but now we can't reproduce it, it works without it and legacy in-memory don't have consistency checks on write operations too. Given all this, I don't think we have to enhance checks here.

**Comment by nvamelichev** [comment]
> legacy in-memory don't have consistency checks on write operations too

This is a good argument :-)

I've looked at the code once again. Now I see that throwing TLI from writes will be an optimization, anyway: `commit()` will throw TLI if the data you wrote to in transaction has been changed by another transaction.

**Pull request merged**